packages/in_app_purchase/in_app_purchase_android/lib/src/billing_client_wrappers/billing_client_manager.dart

```diff
@@ -0,0 +1,156 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+
+import 'package:flutter/widgets.dart';
+
+import 'billing_client_wrapper.dart';
+import 'purchase_wrapper.dart';
+
+/// Abstraction of result of [BillingClient] operation that includes
+/// a [BillingResponse].
+abstract class HasBillingResponse {
+  /// The status of the operation.
+  abstract final BillingResponse responseCode;
+}
+
+/// Utility class that manages a [BillingClient] connection.
+///
+/// Connection is initialized on creation of [BillingClientManager].
+/// If [BillingClient] sends `onBillingServiceDisconnected` event or any
+/// operation returns [BillingResponse.serviceDisconnected], connection is
+/// re-initialized.
+///
+/// [BillingClient] instance is not exposed directly. It can be accessed via
+/// [run] and [runRaw] methods that handle the connection management.
+///
+/// Consider calling [dispose] after the [BillingClient] is no longer needed.
+class BillingClientManager {
+  /// Creates the [BillingClientManager].
+  ///
+  /// Immediately initializes connection to the underlying [BillingClient].
+  BillingClientManager() {
+    _connect();
+  }
+
+  /// Stream of `onPurchasesUpdated` events from the [BillingClient].
+  ///
+  /// This is a broadcast stream, so it can be listened to multiple times.
+  /// A "done" event will be sent after [dispose] is called.
+  late final Stream<PurchasesResultWrapper> purchasesUpdatedStream =
+      _purchasesUpdatedController.stream;
+
+  /// [BillingClient] instance managed by this [BillingClientManager].
+  ///
+  /// In order to access the [BillingClient], consider using [run] and [runRaw]
+  /// methods.
+  @visibleForTesting
+  late final BillingClient client = BillingClient(_onPurchasesUpdated);
+
+  final StreamController<PurchasesResultWrapper> _purchasesUpdatedController =
+      StreamController<PurchasesResultWrapper>.broadcast();
+
+  bool _isConnecting = false;
+  bool _isDisposed = false;
+
+  // Initialized immediately in the constructor, so it's always safe to access.
+  late Future<void> _readyFuture;
```

Comment:
nit: This is optional, but a `Completer<void>` could be used instead of a `Future<void>`. This would make it so you would not have to depend on a late initializer.

Reply:
In this case this wouldn't make much difference - new `Completer` would always have to be created in `_connect` function anyway.